fabfile.py

```diff
@@ -10,65 +10,120 @@
+            path = dirname(__file__)
+            if env.host_string != 'localhost':
+                path = env.get('projectpath', path)
+            fname = join(path, fname[2:])
+        else:
+            fname = join(dirname(rc_filename), fname)
+        service_config = combine_rc(fname, service_config)
+    if overlay is not None:
+        service_config.update(overlay)
+    service_config.pop('_extends', None)
+    service_config.pop('', None)
+    return service_config
 
 
-load_service_configs()
+def filter_global_names(rc_data):
+    """Some keys in rc files are prefixed with * or _ for ini conversion purposes"""
+    return {k.lstrip('*').lstrip('_'): v for (k, v) in rc_data.iteritems()
+            if v != '__delete_key__'}
+
+
+def as_bool(b):
+    return str(b).lower() in {"1", "true", "yes", "t", "on"}
+
+
+def sanitize_env():
+    """Ensure boolean and list env variables are such"""
+    for name in (
+            "uses_memcache", "uses_uwsgi", "uses_apache",
+            "uses_global_supervisor", "uses_apache",
+            "uses_ngnix", "mac", "is_production_env"):
+        # Note that we use as_bool() instead of bool(),
+        # so that a variable valued "False" in the .ini
+        # file is recognized as boolean False
+        setattr(env, name, as_bool(getattr(env, name, False)))
+    public_hostname = env.get("public_hostname", "localhost")
+    if not env.get('hosts', None):
+        env.hosts = [public_hostname]
+    elif not isinstance(env.hosts, list):
+        env.hosts = env.hosts.split()
+    # Note: normally, fab would set host_string from hosts.
+    # But since we use the private name _hosts, and fallback
+    # at this stage within task execution, neither env.hosts
+    # nor env.host_string are set properly. Revisit with Fabric2.
+    if not env.get('host_string', None):
+        env.host_string = env.hosts[0]
+    #Are we on localhost
+    if set(env.hosts) - set(['localhost', '127.0.0.1']) == set():
+        #WARNING:  This code will run locally, NOT on the remote server,
+        # so it's only valid if we are connecting to localhost
+        env.mac = system().startswith('Darwin')
+    else:
+        env.mac = False
+    env.projectpath = env.get('projectpath', dirname(__file__))
+    env.venvpath = env.get('venvpath', join(env.projectpath, 'venv'))
+    env.random_file = env.get('random_file', 'random.ini')
+    env.dbdumps_dir = env.get('dbdumps_dir', join(
+        env.projectpath, '%s_dumps' % env.get("projectname", 'assembl')))
+
+
+def load_rcfile_config():
+    """Load the enviroment from the .rc file."""
+    from fabric.state import env
+    rc_file = env['rcfile']
+    if not rc_file:
+        abort("You must specify a .rc file")
+    if not os.path.exists(env.rcfile):
+        abort("This .rc file does not exist locally: " + rc_file)
+    env.update(filter_global_names(env))
+    env.update(filter_global_names(combine_rc(rc_file)))
+    sanitize_env()
+
+
+def task(func):
```

Comment:
Smart. Can you further elaborate the use of `@wraps`? What is its purpose when you are adding a wrapper yourself?